cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -0,0 +1,528 @@
+/*
+ * Copyright (c) 2025, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.query
+
+import de.fraunhofer.aisec.cpg.graph.*
+import de.fraunhofer.aisec.cpg.graph.AccessValues
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Literal
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.Reference
+
+/**
+ * Determines how far we want to follow edges within the analysis. [maxSteps] defines the total
+ * number of steps we will follow in the graph (unlimited depth if `null`).
+ */
+sealed class AnalysisScope(val maxSteps: Int? = null)
+
+/**
+ * Only intraprocedural analysis. [maxSteps] defines the total number of steps we will follow in the
+ * graph (unlimited depth if `null`).
+ */
+class INTRAPROCEDURAL(maxSteps: Int? = null) : AnalysisScope(maxSteps)
+
+/**
+ * Enable interprocedural analysis. [maxCallDepth] defines how many function calls we follow at most
+ * (unlimited depth if `null`). [maxSteps] defines the total number of steps we will follow in the
+ * graph (unlimited depth if `null`).
+ */
+class INTERPROCEDURAL(val maxCallDepth: Int? = null, maxSteps: Int? = null) :
+    AnalysisScope(maxSteps)
+
+/** Determines in which direction we follow the edges. */
+enum class AnalysisDirection {
```

Comment:
I wonder whether it would make sense to make this a sealed class as well and then make the enum entries objects instead. And then maybe (if that makes sense) we could move some of the analysis logic actually into these Classes instead of having a lot of "if this direction, then" logic in the followX functions.

Reply:
I think we cannot easily remove code from the followX functions (yet) because each subgraph works quite different. But I think that at least some steps can be moved to these classes. E.g. the max depth checks should work already. For others, I'm less confident that it would work right now mainly because the DFG and EOG work quite differently.

I made this also a sealed class now to prepare such a change in this interface.